great_expectations/core/batch_manager.py

```diff
@@ -0,0 +1,169 @@
+import logging
+from collections import OrderedDict
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from great_expectations.core.batch import (
+    Batch,
+    BatchData,
+    BatchDefinition,
+    BatchMarkers,
+)
+from great_expectations.core.id_dict import BatchSpec
+
+if TYPE_CHECKING:
+    from great_expectations.execution_engine import ExecutionEngine
+
+logger = logging.getLogger(__name__)
+logging.captureWarnings(True)
+
+
+class BatchManager:
+    def __init__(
+        self,
+        execution_engine: ExecutionEngine,
+        batch_list: Optional[List[Batch]] = None,
+    ) -> None:
+        """
+        Args:
+            execution_engine: The ExecutionEngine to be used to access cache of loaded Batch objects.
+            batch_list: List of Batch objects available from external source (default is None).
+        """
+        self._execution_engine = execution_engine
+
+        self._active_batch_id = None
+        self._active_batch_data_id = None
+
+        if batch_list is None:
+            batch_list = []
+
+        self._batch_cache = {}
+        self._batch_data_cache = {}
+
+        self.load_batch_list(batch_list=batch_list)
+        if len(batch_list) > 1:
+            logger.info(
+                f"{len(batch_list)} batches will be added to this Validator.  The batch_identifiers for the active "
+                f"batch are {self.active_batch.batch_definition['batch_identifiers'].items()}"
+            )
+
+    @property
+    def execution_engine(self) -> ExecutionEngine:
+        return self._execution_engine
+
+    @property
+    def batch_data_cache(self) -> Dict[str, BatchData]:
+        """Dictionary of loaded BatchData objects."""
+        return self._batch_data_cache
+
+    @property
+    def loaded_batch_ids(self) -> List[str]:
+        """IDs of loaded BatchData objects."""
+        return list(self._batch_data_cache.keys())
+
+    @property
+    def active_batch_data_id(self) -> Optional[str]:
+        """The batch id for the default batch data.
+
+        When a specific Batch objec is unavailable, then the data associated with the active_batch_data_id will be used.
+        """
+        if self._active_batch_data_id is not None:
+            return self._active_batch_data_id
+
+        if len(self._batch_data_cache) == 1:
+            return list(self._batch_data_cache.keys())[0]
```

Comment:
Can we set this batch to the active one and then return its id? It seems like this is getting promoted to the active batch but we are leaving the `_active_batch_data_id` as `None`.

Reply:
@billdirks This is an old "safety fuse" -- if the `_active_batch_data_id` is unavailable (did not get set for whatever reason), then if there is only one loaded `BatchData` (one unambiguously), then it will play the role of an `active_batch_data_id` -- some non-NULL value is needed to compute a metric (by the particular `ExecutionEngine`), and we have been using it.  But if there is more than one, we cannot pinpoint the active `BatchData`, in which case we give up.  Thanks.